src/Client.ts

```diff
@@ -665,32 +666,37 @@ export class Client {
         this.messageParser.read(data);
       }
     };
-    const socketClose = () => {
-      this.connected = false;
-      if (this.socket) {
-        this.socket.removeListener('error', socketError);
-        this.socket.removeListener('close', socketClose);
-        this.socket.removeListener('data', socketData);
-        this.socket.removeListener('end', socketEnd);
-        this.socket.removeListener('timeout', socketTimeout);
-      }
-
-      delete this.socket;
-
-      // Clean up any pending messages
-      for (const [key, messageDetails] of Object.entries(this.messageDetailsByMessageId)) {
-        if (messageDetails.message instanceof UnbindRequest) {
-          // Consider unbind as success since the connection is closed.
-          messageDetails.resolve();
+    const socketClose = ((socketToClose) => {
+      return () => {
+        socketToClose.removeListener('error', socketError);
+        socketToClose.removeListener('close', socketClose);
+        socketToClose.removeListener('data', socketData);
+        socketToClose.removeListener('end', socketEnd);
+        socketToClose.removeListener('timeout', socketTimeout);
+
+        if (this.socket === socketToClose) {
+          this.connected = false;
+          delete this.socket;
         } else {
```

Comment:
This else clause was just for testing to see if I could repro the case and log it.